internetarchive/openlibrary
# Manual reading log check-ins

Closes #6662


Displays the last read date beneath the Want to Read dropper on book pages. If no last read date exists, displays a prompt for last read date when "Already Read" is selected.

Prompt for read date has two options: "Today" and "Custom Date".  If "Today" is clicked, the current date is POSTed. Otherwise, a date picker modal will be displayed.

Dates in the modal are chosen via `select` elements.  If no year is selected, the month and day `select`s are disabled.  Choosing a year enables the month `select`.  Choosing a month adjusts the available day options, and enables the day `select`.
Leap years are handled properly.  There is also a "Today" option that will set the selects to the current date.

Submit button in the modal form is disabled until a year is selected.

### Technical

This is only available to beta-testers.

### Testing

While logged in as a `beta-tester`:
1. Navigate to a book page for a book that is not on any reading log shelf.
2. Select "Already Read" from the reading log dropper.
3. Ensure that the date prompt appears below the dropper.
4. Click the "Today" link.  The prompt should be replaced with a component displaying the last read date.  Refresh the page, and ensure that the last read date is displayed.
5. Navigate to another edition of the book. Ensure that the date prompt appears below the "Already Read" button.
6. Click the "Custom Date" link.  A modal containing a date selection form should open. Ensure that none of the date select elements are populated, and that the following elements are disabled: submission button, month select, day select.
7. Click outside of the modal and ensure that it closes.  Open the modal again by clicking "Custom Date"
8. Ensure that the correct amount of day `options` are present for the given month. Change the month and check again.
9. Change month to the default value. Ensure that the day `select` element is disabled.  Change the month back.
10. Change year to the default value.  Ensure that the submit button, and the day and month `select` elements are disabled.
11. Change the year to a leap year, and the month to February.  Ensure that there are 29 days available to select.
12. Click the "Today" link in the modal.  Ensure that today's date is populated in the select elements.
13. Submit a check-in date.  Ensure that the sidebar updates with your check-in date.
14. Click the "Edit Date" link.  Ensure that the date modal opens and is populated with the last check-in date.
15. Change the date and submit the form.  Ensure that the new date is displayed in the sidebar.

### Screenshot

![enable_disable](https://user-images.githubusercontent.com/28732543/199557933-0df1d5dc-7125-44e4-928a-a70229d08f1d.gif)
_Elements are disabled when appropriate._

![Screenshot from 2022-11-02 10-19-16](https://user-images.githubusercontent.com/28732543/199558274-2622e103-9b6b-4d3f-96ba-21fded5c1d61.png)
_The modal on smaller screens._
### Stakeholders
